src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c

```diff
@@ -0,0 +1,393 @@
+  SODERLIND_K4(C) = RCONST(0.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create a PI controller (subset of Soderlind)
+ */
+
+SUNAdaptController SUNAdaptController_PI(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_PI(C,
+                                         DEFAULT_PI_K1,
+                                         DEFAULT_PI_K2);
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set PI parameters
+ */
+
+int SUNAdaptController_SetParams_PI(SUNAdaptController C,
+                                    sunrealtype k1, sunrealtype k2)
+{
+  SODERLIND_K1(C) = k1;
+  SODERLIND_K2(C) = k2;
+  SODERLIND_K3(C) = RCONST(0.0);
+  SODERLIND_K4(C) = RCONST(0.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create an I controller (subset of Soderlind)
+ */
+
+SUNAdaptController SUNAdaptController_I(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_I(C, DEFAULT_I_K1);
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set PI parameters
+ */
+
+int SUNAdaptController_SetParams_I(SUNAdaptController C, sunrealtype k1)
+{
+  SODERLIND_K1(C) = k1;
+  SODERLIND_K2(C) = RCONST(0.0);
+  SODERLIND_K3(C) = RCONST(0.0);
+  SODERLIND_K4(C) = RCONST(0.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create an explicit Gustafsson controller
+ */
+
+SUNAdaptController SUNAdaptController_ExpGus(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_ExpGus(C,
+                                             DEFAULT_EXPGUS_K1,
+                                             DEFAULT_EXPGUS_K2);
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set explicit Gustafsson parameters
+ */
+
+int SUNAdaptController_SetParams_ExpGus(SUNAdaptController C,
+                                        sunrealtype k1, sunrealtype k2)
+{
+  SODERLIND_K1(C) = k1+k2;
+  SODERLIND_K2(C) = -k2;
+  SODERLIND_K3(C) = RCONST(0.0);
+  SODERLIND_K4(C) = RCONST(0.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create an implicit Gustafsson controller
+ */
+
+SUNAdaptController SUNAdaptController_ImpGus(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_ImpGus(C,
```

Comment:
It unlikely but we should check this return value, destroy the controller, and return `NULL` if setting the parameters (somehow) fails.

Reply:
See comments below about checking for a `NULL` input